### PR TITLE
fix #6679

### LINKED ISSUE #6679: Js_of_ocaml separate compilation issue when using some jsoo flags.
Users can pass arbitrary flags to js_of_ocaml (since https://github.com/ocaml/dune/pull/5049).
When using separate compilation, js_of_ocaml needs to compile multiple bits (runtime.js, deps*.cma.js, and *cmo.js) separately and link everything together.

The isssue is that some jsoo flags (e.g `--enable use-js-string`, `--enable effects`) require that all individual bits are compiled the same way or the generated js might not work correctly. https://github.com/ocsigen/js_of_ocaml/pull/1346 improves the situation a bit by having jsoo complain if this is not the case, instead of emitting invalid JS.

As far as I know, we mostly have an issue with how we compile cma files to js.
We currently setup a single rule per lib (in `_build/default/.js/{LIB}/` for findlib libraries and in `_build/default/{PATH_TO_LIB}/.{LIB}.objs/`) but we would need to generate rules based on the jsoo flags used by executables.

One easy solution would be to declare all cma.js rules needed by an executable for each executables but we would lose sharing which kind of defeat the purpose of separate compilation.

Any suggestion ?

### PR DESCRIPTION
**Pull request by hhugo**
Attempt to fix #6679 
and cleanup the jsoo rules

### CODE REVIEW

File: boot/libs.ml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 let executables = [ "main" ]
 
-let external_libraries = [ "unix"; "threads.posix" ]
+let external_libraries = [ "unix"; "threads"; "threads.posix" ]
```
Comment (rgrinberg) [inline boot/libs.ml:3]: Did you commit this accidentally? I think this change needs to be reverted and everything should work in CI.

File: test/blackbox-tests/test-cases/jsoo/no-check-prim.t/run.t [hunk 2]
```diff
@@ -56,7 +56,6 @@ Compilation using jsoo
   >   (default (disable_dynamically_linked_foreign_archives true)))
   > EOF
   $ dune build --display short bin/technologic.bc.js @install --profile dev
-   js_of_ocaml bin/technologic.bc.runtime.js
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/jsoo/no-check-prim.t/run.t:59]: How come the runtime isn't built here anymore?

**Comment by rgrinberg** [comment]
The PR looks good to me. There's quite a few unrelated changes that I would appreciate if they were moved to separate commits, but it's not strictly necessary.

The issue of setting up rules for every library lazily requires rule streaming. That's something we're going to tackle early 2023 (I hope).

File: boot/libs.ml (hunk 1 above)
Comment (hhugo) [reply boot/libs.ml:3]: I don't know how this got here, will revert

Comment (Alizter) [reply boot/libs.ml:3]: Something similar happened to me also. When you do make bootstrap or make dune.exe.

Comment (hhugo) [reply boot/libs.ml]: now reverted

**Comment by hhugo** [comment]
> The PR looks good to me. There's quite a few unrelated changes that I would appreciate if they were moved to separate commits, but it's not strictly necessary.
> 
> The issue of setting up rules for every library lazily requires rule streaming. That's something we're going to tackle early 2023 (I hope).

I've split this PR in 7 commits that can be reviewed individually.

File: test/blackbox-tests/test-cases/jsoo/no-check-prim.t/run.t (hunk 2 above)
Comment (hhugo) [reply test/blackbox-tests/test-cases/jsoo/no-check-prim.t/run.t:59]: I don't know why it was rebuild here. It was already build on line 5

**Review by rgrinberg — APPROVED** [review]
LGTM.

Does this need a CHANGES entry? I can't tell what's user facing about this change.

**Comment by hhugo** [comment]
> LGTM.
> 
> Does this need a CHANGES entry? I can't tell what's user facing about this change.

This PR mostly changes where intermediate artifact are stored. The user facing change is that one can now use `--enable (effects|use-js-string)` and get back a working js file when using separate compilation.
I'll try to add a changelog entry. It always help when there is an unexpected change of behavior.

**Comment by hhugo** [comment]
I've added a changelog entry

**Pull request merged**